File: scripts/merge_unit1_pdf.js
const fs = require('fs');
const path = require('path');

function read(p) {
  try { return fs.readFileSync(p, 'utf8'); } catch(e) { return ''; }
}

function splitBlocks(text) {
  // Split on lines that start with a number and a dot
  const parts = text.split(/(?=^\d+\.)/m);
  return parts.map(p => p.trim()).filter(Boolean);
}

function normalizeForKey(block) {
  const lines = block.split(/\r?\n/).map(l => l.trim()).filter(Boolean);
  const keep = [];
  for (let l of lines) {
    if (/^(ANSWER:|Source:|Source|\-\-+)/i.test(l)) continue;
    if (/^[A-D]\)|^\([A-D]\)/.test(l)) continue;
    if (/^\d+\.$/.test(l)) continue;
    keep.push(l);
  }
  let text = keep.join(' ');
  text = text.replace(/[^\w\s]+/g, ' ').toLowerCase().trim();
  return text.slice(0, 180);
}

function main(){
  const a = path.join('public','questions','unit-1.txt');
  const b = path.join('tmp','pdfs','unit-1-guide.txt');
  const out = path.join('public','questions','unit-1-merged.txt');
  const bak = path.join('public','questions','unit-1.backup.txt');

  const ta = read(a);
  const tb = read(b);
  if (!ta && !tb) { console.log('No input files found.'); return; }
  if (ta) fs.writeFileSync(bak, ta, 'utf8');
  if (ta) console.log(`Backed up ${a} -> ${bak}`);

  let blocks = [];
  if (ta) blocks = blocks.concat(splitBlocks(ta));
  if (tb) blocks = blocks.concat(splitBlocks(tb));

  const seen = new Set();
  const merged = [];
  for (let blk of blocks) {
    const key = normalizeForKey(blk);
    if (!key) continue;
    if (seen.has(key)) continue;
    seen.add(key);
    merged.push(blk);
  }

  const outLines = [];
  outLines.push('AP WORLD HISTORY - UNIT 1 MERGED QUESTIONS');
  outLines.push(`Total unique questions: ${merged.length}`);
  outLines.push('');
  merged.forEach((blk, i) => {
    const blk2 = blk.replace(/^\s*\d+\./m, '').trim();
    outLines.push(`${i+1}. ${blk2}`);
    outLines.push('\n');
  });

  fs.writeFileSync(out, outLines.join('\n'), 'utf8');
  console.log(`Wrote merged file: ${out} (questions=${merged.length})`);
}

main();

File: scripts/merge_unit1_pdf.py
#!/usr/bin/env python3
import re
from pathlib import Path

def read_file(p):
    return Path(p).read_text(encoding='utf-8') if Path(p).exists() else ''

def split_blocks(text):
    # Find numbered question blocks like '1.' at line starts
    parts = re.split(r'(?m)^(?=\d+\.)', text)
    blocks = [p.strip() for p in parts if p.strip()]
    return blocks

def normalize_for_key(block):
    # Remove ANSWER lines, SOURCE lines, separators, and choices
    lines = []
    for line in block.splitlines():
        l = line.strip()
        if not l:
            continue
        if re.match(r'^(ANSWER:|Source:|Source|--+)', l, re.I):
            continue
        # skip choice lines like A) or (A)
        if re.match(r'^[A-D]\)|^\([A-D]\)', l):
            continue
        # skip lines that are just numbers
        if re.match(r'^\d+\.$', l):
            continue
        lines.append(l)
    text = ' '.join(lines)
    # remove punctuation, lowercase
    text = re.sub(r'[\W_]+', ' ', text).lower().strip()
    return text[:180]

def main():
    a = 'public/questions/unit-1.txt'
    b = 'tmp/pdfs/unit-1-guide.txt'
    out = 'public/questions/unit-1-merged.txt'
    bak = 'public/questions/unit-1.backup.txt'

    ta = read_file(a)
    tb = read_file(b)

    if not ta and not tb:
        print('No input files found.')
        return

    # Backup existing file
    if ta:
        Path(bak).write_text(ta, encoding='utf-8')
        print(f'Backed up {a} -> {bak}')

    blocks = []
    if ta:
        blocks += split_blocks(ta)
    if tb:
        blocks += split_blocks(tb)

    seen = set()
    merged = []
    for blk in blocks:
        key = normalize_for_key(blk)
        if not key:
            continue
        if key in seen:
            continue
        seen.add(key)
        merged.append(blk)

    # Write merged with numbering
    out_lines = []
    out_lines.append('AP WORLD HISTORY - UNIT 1 MERGED QUESTIONS')
    out_lines.append(f'Total unique questions: {len(merged)}')
    out_lines.append('')
    for i, blk in enumerate(merged, 1):
        # Remove original numbering if present, then prepend new number
        blk2 = re.sub(r'(?m)^\s*\d+\.', '', blk).strip()
        out_lines.append(f'{i}. {blk2}')
        out_lines.append('\n')

    Path(out).write_text('\n'.join(out_lines), encoding='utf-8')
    print(f'Wrote merged file: {out} (questions={len(merged)})')

if __name__ == '__main__':
    main()
